add alternative ways to specify ORF1a or ORF1b

anastasia-escher opened the pull request:
resolves #496 (probably)

anastasia-escher committed:
add alternative ways to specify ORF1a or ORF1b

chaoran-chen commented:
I don't think that it is already fully working. In the following example, it is not translating?

![image](https://user-images.githubusercontent.com/18666552/169394211-79d7dbe5-ae3a-4bf9-a151-9dca6dd5eb71.png)

anastasia-escher commented:
Oh, it works now only with lower case - nsp, so I will make it case-independent and recommit .

anastasia-escher commented:
Hm, but now I see that it works also with upper case for me, strange. 
![Screenshot_20220519-224904_DuckDuckGo](https://user-images.githubusercontent.com/52914023/169402258-953f5003-b96d-4f2b-b800-b995a0385dab.jpg)

anastasia-escher commented:
Ah, now I understand, it only shows the alternative in the autocomolete options if you type, but it doesn't translate if you paste. I will change it.

anastasia-escher committed:
fix autocomplete options

anastasia-escher committed:
fix translation

anastasia-escher committed:
fix auto translation

anastasia-escher commented:
@chaoran-chen please check if it works correctly now.

chaoran-chen commented:
@anastasia-escher, it is now translating after pressing enter but it would also be nice if it translates the autosuggestion as well:
![image](https://user-images.githubusercontent.com/18666552/169786255-a1c73eb1-6e36-4fe5-b640-fb44abec54d8.png)

Further, if someone enters a NSP... value, it does not need to show the ORFab equivalent. And the same when someone enters the ORFab value. We only need a translation to ORFa/ORFb.

![image](https://user-images.githubusercontent.com/18666552/169786438-122109d3-b3b1-4a6f-946c-2f96f977dc5b.png)

anastasia-escher committed:
add translation on autocomplete

anastasia-escher commented:
@chaoran-chen please check if it works correctly now.

chaoran-chen commented:
Sorry if I was still unclear: When the user enters "NSP...", then it should show something like "ORF... (= NSP...)". When the user enters "ORF1ab ...", it should show "ORF... (= ORFab...)"

anastasia-escher committed:
fix autocomplete

anastasia-escher commented:
@chaoran-chen and now?

chaoran-chen commented:
Still not getting an auto suggestion in every case? (First issue mentioned in https://github.com/cevo-public/cov-spectrum-website/pull/502#issuecomment-1134406011)

![image](https://user-images.githubusercontent.com/18666552/171584946-74007041-6151-4f64-b0d0-8bd8a13140aa.png)

anastasia-escher committed:
suggestion for every case

anastasia-escher commented:
@chaoran-chen now it should work

corneliusroemer commented:
Tested it a bit and it seems to work well :)

corneliusroemer commented:
@anastasia-escher you really need to start with the keywords: `resolves` `fixes` etc, if you add `probably` before Github won't link.

You don't need to be certain that it fixes it, it's enough that that's what you're trying to achieve.

So I edited your PR message to `resolves #496 (probably)` so now it's linked :)

chaoran-chen commented:
If I enter `NSP4:4K`, it gets translated to `ORF1a:3767`. The `K` got lost.

anastasia-escher commented:
> @anastasia-escher you really need to start with the keywords: `resolves` `fixes` etc, if you add `probably` before Github won't link.
> 
> You don't need to be certain that it fixes it, it's enough that that's what you're trying to achieve.
> 
> So I edited your PR message to `resolves #496 (probably)` so now it's linked :)

oh sorry, I always forget about it

anastasia-escher committed:
fix issue with the letter at the end

chaoran-chen committed:
Merge branch 'develop' into autocomplete

chaoran-chen commented:
![image](https://user-images.githubusercontent.com/18666552/172070927-58d566a9-9426-4433-ba4e-774fa04e1816.png)


@anastasia-escher, there is no need to show the `ORF1ab`-equivalence if the user enters a mutation using our main notation (i.e., ORF1a or ORF1b). Just to clarify again, we only show translations in one direction: from the NSP notation and the ORF1ab notation to our notation, not back.

anastasia-escher committed:
fix autocomplete for the main notation

anastasia-escher committed:
fix autocomplete for the main notation

anastasia-escher committed:
fix autocomplete for the mail notation

chaoran-chen commented:
Another small inconsistency/strange behavior: When entering a NSP value, the (optional) character before the position disappears:

![image](https://user-images.githubusercontent.com/18666552/172494247-24011f19-4e15-4f5b-b90b-0d1d91bb17fc.png)

This does not happen when entering an ORF1ab value:

![image](https://user-images.githubusercontent.com/18666552/172494291-59afdfd0-5d22-450c-8f58-87d2f110270b.png)

anastasia-escher committed:
fix alternative letter in nsp notation

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+  nsp16: 6799,
+};
+
+function getEquivalent(value: string, notation: string) {
```
- It would be better to narrow down the type of `notation`: I think that it could be `'nsp' | 'orf1ab'`?
- I think that this function's logic is not obvious and it would be great if you could document/explain the algorithm a bit.

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+function getEquivalent(value: string, notation: string) {
+  let orf1aorbcodon: number | null = null;
+  let nsp: string | null = null;
+  let nspCodon: number | null = null;
```
Is there a reason why you initiate the three variables above already here and not when they are needed?
In which case do they need to be `null`?

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+  let orf1aorbcodon: number | null = null;
+  let nsp: string | null = null;
+  let nspCodon: number | null = null;
+  let combinedCodon: number | null = 3646;
```
In which case could `combinedCodon` be `null`?

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+  }
+}
+
+const translateMutation = (oldValue: string) => {
```
This function shares some code with `getEquivalent()`. Is it possible to reduce code redundancy?

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -101,36 +248,37 @@ export const VariantSearchField = ({ onVariantSelect, currentSelection, triggerS
+    let options: string[] = [
       'S:D614G',
-      'ORF1b:P314L',
+      `ORF1b:P314L`,
```
To keep the PR clean, please try to avoid unnecessary edits and revert the changes from ' to `

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+  let combinedCodon: number | null = 3646;
+  let mutationArray = oldValue.split(':');
+
+  if (mutationArray[0] === 'ORF1b' || mutationArray[0] === 'ORF1a') {
```
Is this part of the function outdated? Since we don't need to translate ORF1a/ORF1b mutations into another notation?

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+    }
+
+    nsp = mutationArray[0];
+    combinedCodon = nsps[nsp] + nspCodon - 1;
```
The auto-suggestion should show nothing if the user types a non-existing NSP and should not translate it to `ORF1a:NaN`:

![image](https://user-images.githubusercontent.com/18666552/174026264-dd326283-57e0-40d3-8e1e-dc862b8b0ae2.png)

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -25,10 +25,157 @@ const backgroundColor: { [key in SearchType]: string } = {
+    actualValue = optionString.split(' ')[0];
+  } else if (optionString.toLowerCase().startsWith('nsp')) {
+    if (translateMutation(optionString.toLocaleLowerCase())) {
+      actualValue = translateMutation(optionString.toLocaleLowerCase());
```
Is it intended to use `toLocaleLowerCase()` instead of `toLowerCase()` like everywhere else here?

chaoran-chen reviewed (changes requested):
Hi @anastasia-escher, I'm sorry for the slow response times! I left a number of comments in the code. In addition, the parsing of comma and plus-sign separated lists of values does not work yet.

If I paste `nsp6:77,ORF1ab:A500G,ORF1b:34,nsp14:320T` into the search field, only ORFab:34 gets parsed but the others are not getting translated.

anastasia-escher committed:
add multiple values input for nsp and orf1ab notation; remove code redundancy

chaoran-chen committed:
Correct NSP8 position

Following the change in https://github.com/theosanderson/Codon2Nucleotide/commit/ebfb8e2a9cf28bc66206837c06e03e67da212263

chaoran-chen reviewed (approved):
Thank you @anastasia-escher, this looks great!
